File: answers/lintcode/Triangle.cpp
class Solution {
public:
    /**
     * @param triangle: a list of lists of integers.
     * @return: An integer, minimum path sum.
     */
    int minimumTotal(vector<vector<int> > &triangle) {
        // write your code here
		for (size_t i = 1; i < triangle.size(); ++ i) {
			for (size_t j = 0; j < triangle[i].size(); ++ j) {
				triangle[i][j] += min((j > 0 ? triangle[i - 1][j - 1] : INT_MAX), (j < triangle[i - 1].size() ? triangle[i - 1][j] : INT_MAX));
				//cout << "test " << i << " " << j << " " << triangle[i][j] << endl;
			}
		}
		int min_total = INT_MAX;
		for (size_t i = 0; i < triangle.back().size(); ++ i) {
			min_total = min(min_total, triangle.back()[i]);
		}
		return min_total;
    }
};

File: answers/leetcode/Triangle/Triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        List<Integer> total = new ArrayList<>();
        for (int i = 0; i < triangle.size(); i++) {
            List<Integer> newTotal = new ArrayList<>();
            for (int j = 0; j < triangle.get(i).size(); j++) {
                int sum = Integer.MAX_VALUE;
                if (j < total.size()) {
                    sum = Math.min(sum, triangle.get(i).get(j) + total.get(j));
                }
                if (j > 0) {
                    sum = Math.min(sum, triangle.get(i).get(j) + total.get(j - 1));
                }
                if (sum == Integer.MAX_VALUE) {
                    sum = triangle.get(i).get(j);
                }
                newTotal.add(sum);
            }
            total = newTotal;
        }
        int minTotal = Integer.MAX_VALUE;
        for (int sum : total) {
            minTotal = Math.min(minTotal, sum);
        }
        return minTotal;
    }
}